Zend/zend_hash.c

```diff
@@ -788,6 +790,10 @@ static zend_always_inline zval *_zend_hash_add_or_update_i(HashTable *ht, zend_s
 				ht->pDestructor(data);
 			}
 			ZVAL_COPY_VALUE(data, pData);
+			if (!(GC_FLAGS(ht) & IS_ARRAY_PERSISTENT)
+				&& (Z_TYPE_P(data) == IS_REFERENCE || Z_TYPE_P(data) == IS_OBJECT)) {
```

Comment:
It make sense to check GC_NOT_COLLECTABLE instead of IS_ARRAY_PERSISTENT. This will at least shorten the check for already collectable arrays.

Reply:
I think we need both, actually, because we shouldn't add the `GC_NOT_COLLECTABLE` flag to persistent arrays.

Reply:
Why? Persistent arrays can't be freed, so they never should be collected by GC. Currently, we always set ``GC_NOT_COLLECTABLE`` together with ``GC_PERSISTENT``.

Reply:
`gc_ht_check_collectable` removes the `GC_NOT_COLLECTABLE` flag, which should not happen for persistent arrays as they are never collectable, right?

Reply:
PHP should never modify ``GC_PERSISTENT`` arrays in-place. So PHP will never reset ``GC_NOT_COLLECTABLE`` flag of ``GC_PERSISTENT`` array. Therefore the single check for ``GC_NOT_COLLECTABLE`` should be  enough.